File: kotlin-node/karakum/src/jsMain/resources/plugins/convertWebCryptoQualifiedName.js
import ts from "typescript";

export default function (node, context, render) {
    if (
        ts.isQualifiedName(node)
        && ts.isIdentifier(node.left)
        && node.left.text === "webcrypto"
    ) {
        if (node.right.text === "BufferSource") {
            return "js.buffer.BufferSource"
        }

        return `web.crypto.${render(node.right)}`
    }

    if (
        ts.isPropertyAccessExpression(node)
        && ts.isIdentifier(node.expression)
        && node.expression.text === "webcrypto"
    ) {
        return `web.crypto.${render(node.name)}`
    }

    return null
}

File: kotlin-node/karakum/src/jsMain/kotlin/node/karakum/plugins/convertWebCryptoQualifiedName.kt
package node.karakum.plugins

import node.karakum.util.nullable
import io.github.sgrishchenko.karakum.extension.createPlugin
import typescript.isIdentifier
import typescript.isPropertyAccessExpression
import typescript.isQualifiedName

val convertWebCryptoQualifiedName = createPlugin { node, context, render ->
    nullable {
        ensure(isQualifiedName(node))

        val left = node.left
        ensure(isIdentifier(left))
        ensure(left.text == "webcrypto")

        nullable {
            ensure(node.right.text == "BufferSource")

            "js.buffer.BufferSource"
        } ?: run {
            "web.crypto.${render(node.right)}"
        }
    } ?: nullable {
        ensure(isPropertyAccessExpression(node))

        val expression = node.expression
        ensure(isIdentifier(expression))
        ensure(expression.text == "webcrypto")

        "web.crypto.${render(node.name)}"
    }
}
